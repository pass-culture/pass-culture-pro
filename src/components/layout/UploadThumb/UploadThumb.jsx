import classnames from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import AvatarEditor from 'react-avatar-editor'
import Dropzone from 'react-dropzone'

const IMAGE_MAX_SIZE = 10 // in MB

export function computeNewZoom(current, min, max, step, factor, direction) {
  const zoom = current + step * factor * direction

  if (zoom >= min && zoom <= max) {
    return zoom
  }

  return current
}

class UploadThumb extends Component {
  constructor() {
    super()
    this.avatarEditor = React.createRef()
    this.handleSetZoomInput = React.createRef()
    this.state = {
      isEdited: false,
      readOnly: false,
      image: null,
      isUploadDisabled: false,
      isDragging: false,
      zoom: 1,
    }
  }

  static getDerivedStateFromProps(props, prevState) {
    const readOnly = props.hasExistingImage && !prevState.isEdited
    return {
      readOnly,
      isDragging: prevState.isDragging,
      image: props.image || prevState.image,
    }
  }

  handleDragStart = () => {
    this.setState({
      dragging: true,
    })
  }

  handleDragStop = () => {
    this.setState({
      dragging: false,
    })
  }

  handleDrop = dropped => {
    const { maxSize } = this.props
    const image = dropped[0]
    // convert into MB
    const size = image.size / 1048576
    this.setState({
      isDragging: false,
      isUploadDisabled: size > maxSize,
      image,
      size,
    })
  }

  handleOnZoomChange = event => {
    this.setState({ zoom: parseFloat(event.target.value) })
  }

  handleOnImageChange = ctx => {
    console.log('CTX', ctx)
    const { image, isUploadDisabled } = this.state
    if (!image) return
    const { onImageChange } = this.props
    if (onImageChange) {
      if (isUploadDisabled) return onImageChange(ctx)

      onImageChange(ctx, image, this.avatarEditor.current.getCroppingRect())
    }
  }

  changeZoom(direction) {
    const { zoom } = this.state

    const factor = 10 // Slider step is too low for button usage
    const step = parseFloat(this.handleSetZoomInput.current.getAttribute('step'))
    const min = parseFloat(this.handleSetZoomInput.current.getAttribute('min'))
    const max = parseFloat(this.handleSetZoomInput.current.getAttribute('max'))

    const newZoom = computeNewZoom(zoom, min, max, step, factor, direction)

    this.setState({ zoom: newZoom })
  }

  handleOnChangeImageClick = () => this.setState({ isEdited: true })

  handleIncrement = () => this.changeZoom(1)

  handleDecrement = () => this.changeZoom(-1)

  render() {
    const { border, borderRadius, className, height, maxSize, width } = this.props
    const { dragging, image, isUploadDisabled, readOnly, size, zoom } = this.state
    return (
      <div className="field">
        <div className={classnames('upload-thumb', className)}>
          <Dropzone
            className={classnames('dropzone', {
              'has-image': Boolean(image),
              'no-drag': readOnly,
            })}
            disableClick={Boolean(image || readOnly)}
            onDragEnter={this.handleDragStart}
            onDragLeave={this.handleDragStop}
            onDrop={this.handleDrop}
          >
            {image && (
              <div
                className={`drag-n-drop ${dragging ? 'dragged' : ''}`}
                style={{ borderRadius, width, height }}
              >
                {'Cliquez ou glissez-d√©posez pour charger une image'}
              </div>
            )}
            <AvatarEditor
              border={border}
              borderRadius={borderRadius}
              className="avatar editor"
              color={[255, 255, 255, readOnly || !image ? 1 : 0.6]}
              height={height}
              image={image}
              onImageChange={this.handleOnImageChange}
              ref={this.avatarEditor}
              scale={zoom}
              width={width}
            />
            {!readOnly && image && (
              <div id="zoomControl">
                <button
                  className="change-zoom decrement"
                  onClick={this.handleDecrement}
                  type="button"
                >
                  <span>
                    {'-'}
                  </span>
                </button>
                <input
                  className="zoom level-left"
                  max="4"
                  min="1"
                  name="zoomLeft"
                  onChange={this.handleOnZoomChange}
                  ref={this.handleSetZoomInput}
                  step="0.01"
                  type="range"
                  value={zoom}
                />
                <button
                  className="change-zoom increment"
                  onClick={this.handleIncrement}
                  type="button"
                >
                  <span>
                    {'+'}
                  </span>
                </button>
              </div>
            )}
          </Dropzone>
          <nav className="field content">
            {isUploadDisabled && (
              <p className="has-text-danger">
                {`Votre image trop volumineuse : ${size.toFixed(2)} > ${maxSize}Mo`}
              </p>
            )}
          </nav>
        </div>
      </div>
    )
  }
}

UploadThumb.defaultProps = {
  border: 25,
  borderRadius: 250,
  className: null,
  height: 250,
  maxSize: IMAGE_MAX_SIZE,
  width: 250,
}

UploadThumb.propTypes = {
  border: PropTypes.number,
  borderRadius: PropTypes.number,
  className: PropTypes.string,
  hasExistingImage: PropTypes.bool.isRequired,
  height: PropTypes.number,
  image: PropTypes.oneOfType([PropTypes.string, PropTypes.shape()]).isRequired,
  maxSize: PropTypes.number,
  onImageChange: PropTypes.func.isRequired,
  width: PropTypes.number,
}

export default UploadThumb
