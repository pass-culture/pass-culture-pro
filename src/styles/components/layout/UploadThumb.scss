.upload-thumb {
  position: relative;
  display: block;
  text-align: center;
  height: auto;

  .input {
    height: auto;
  }

  img.read-only {
    box-sizing: content-box;
    object-fit: cover;
    border-style: solid;
    border-color: transparent;
    margin: auto;
  }

  .drag-n-drop {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: $grey-lighter;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 2rem;
    margin: auto;
    cursor: pointer;

    &:before {
      content: '';
      position: absolute;
      border-radius: inherit;
      margin: 1rem;
      border: thin dashed $grey-light;
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
    }

    &.dragged {
      font-weight: $weight-bold;
      outline-color: $text;
      &:before {
        border-color: $text;
      }
    }
  }

  .dropzone {
    background: transparent;
    position: relative;
    margin: auto;
    margin-bottom: 2rem;
    &.no-drag {
      &:before {
        position: absolute;
        top: 0;
        bottom: 0;
        left: 0;
        right: 0;
        content: '';
      }
    }
    &.has-image {
      .drag-n-drop {
        display: none;
      }
    }
    .zoom {
      width: calc(100% - 90px);
      cursor: pointer;
    }

    .change-zoom {
      width: 40px;
      height: 40px;
      margin: 0 0 0 5px;
      display: inline-block;
      border: 2px solid $red;
      border-radius: 50%;
      background: transparent;
      color: $red;
      font-size: 40px;
      padding: 0;

      span {
        display: block;
        margin-top: -9px;
      }
    }

    .change-zoom.decrement {
      margin: 0 5px 0 0;
    }
  }

  input[type=range]::-webkit-slider-thumb {
    background: url(/thumb.png) center no-repeat $red !important;
  }


  canvas {
    display: block;
    margin: auto;
  }

}

#zoomControl {
  display: flex;
  flex: 1;
  width: 68%;
  margin: 10px auto;

  button, input {
    flex-direction: row;

  }
}
