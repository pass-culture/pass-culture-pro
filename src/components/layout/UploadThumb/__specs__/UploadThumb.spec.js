import React from 'react'

import { shallow } from 'enzyme'

import UploadThumb, { computeNewZoom, isImageTooLarge } from '../UploadThumb'

const defaultParams = {
  current: 0.3,
  min: 0.1,
  max: 3,
  step: 0.01,
  factor: 10,
  direction: 1,
}

describe('src | components | layout | UploadThumb |', () => {
  let props
  let image = {
    lastModified: 1565378546222,
    lastModifiedDate: 'Fri Aug 09 2019 21:22:26 GMT+0200 (heure d’été d’Europe centrale)',
    name: 'IMG_4366.jpg',
    size: 1503804,
    type: 'image/jpeg',
    webkitRelativePath: '',
  }

  beforeEach(() => {
    props = {
      collectionName: 'fake',
      dispatch: jest.fn(),
      hasExistingImage: false,
      image,
      onImageChange: jest.fn(),
      storeKey: 'image',
    }
  })

  it('should match the snapshot', () => {
    // given
    props.image = {
      name: 'IMG_4366.jpg',
      size: 1503804,
      type: 'image/jpeg',
    }

    // when
    const wrapper = shallow(<UploadThumb {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('render', () => {
    it('should render main component with default state', () => {
      // when
      const wrapper = shallow(<UploadThumb {...props} />)

      // then
      expect(wrapper.state('readOnly')).toBe(false)
      expect(wrapper.state('image')).toStrictEqual(image)
      expect(wrapper.state('isUploadDisabled')).toBe(false)
      expect(wrapper.state('zoom')).toBe(1)
    })

    describe('editor-zone div', () => {
      it('should render main component properly', () => {
        // given
        const wrapper = shallow(<UploadThumb {...props} />)

        // when
        const editorZoneComponent = wrapper.find('.editor-zone')

        // then
        expect(editorZoneComponent).toHaveLength(1)
      })

      it('should not have no-drag class when is readOnly', () => {
        // given
        props.image = {
          name: 'editor-zoneExample.jpg',
          size: 1503804,
          type: 'image/jpeg',
        }
        props.hasExistingImage = true

        // when
        const wrapper = shallow(<UploadThumb {...props} />)
        const editorZoneComponent = wrapper.find('.editor-zone')

        // then
        expect(editorZoneComponent.props().className).toStrictEqual('editor-zone has-image no-drag')
      })
    })

    describe('image too large alert message', () => {
      it('should not be displayed when image is smaller', () => {
        // when
        const wrapper = shallow(<UploadThumb {...props} />)
        const showAlertComponent = wrapper.find('.has-text-danger')

        // then
        expect(showAlertComponent).toHaveLength(0)
      })

      it('should not be displayed when image is larger', () => {
        // given
        props.image = {
          name: 'editor-zoneExample.jpg',
          size: 150380467,
          type: 'image/jpeg',
        }
        props.hasExistingImage = true

        // when
        const wrapper = shallow(<UploadThumb {...props} />)
        const showAlertComponent = wrapper.find('.has-text-danger')

        // then
        expect(showAlertComponent.text()).toStrictEqual(
          'Votre image trop volumineuse : 150.380467 > 10 Mo'
        )
      })
    })

    describe('zoomcontrol', () => {
      it('should not be displayed when image is uploaded', () => {
        // given
        props.hasExistingImage = true

        // when
        const wrapper = shallow(<UploadThumb {...props} />)

        const zoomControlComponent = wrapper.find('#zoomControl')

        // then
        expect(zoomControlComponent).toHaveLength(0)
      })

      it('should be displayed when image is uploaded', () => {
        // given
        props.hasExistingImage = false

        // when
        const wrapper = shallow(<UploadThumb {...props} />)

        const zoomControlComponent = wrapper.find('#zoomControl')

        // then
        expect(zoomControlComponent).toHaveLength(1)
      })

      it('should contain one button to zoom out', () => {
        // given
        props.hasExistingImage = false

        // when
        const wrapper = shallow(<UploadThumb {...props} />)

        const zoomControlComponent = wrapper
          .find('#zoomControl')
          .find('button')
          .at(0)

        // then
        expect(zoomControlComponent).toHaveLength(1)
        expect(zoomControlComponent.props().className).toStrictEqual('change-zoom decrement')
      })

      it('should contain one button to zoom in', () => {
        // given
        props.hasExistingImage = false

        // when
        const wrapper = shallow(<UploadThumb {...props} />)

        const zoomControlComponent = wrapper
          .find('#zoomControl')
          .find('button')
          .at(1)

        // then
        expect(zoomControlComponent).toHaveLength(1)
        expect(zoomControlComponent.props().className).toStrictEqual('change-zoom increment')
      })
    })
  })

  describe('handleDecrement', () => {
    it('should decrement zoom on click on minus button', () => {
      // given
      const getAttribute = jest.fn().mockImplementation(attribute => {
        if (attribute === 'step') {
          return 0.01
        }
        if (attribute === 'min') {
          return 1
        }
        if (attribute === 'max') {
          return 4
        }
      })

      // when
      const wrapper = shallow(<UploadThumb {...props} />)
      wrapper.instance()['handleSetZoomInput'] = {
        current: {
          getAttribute,
        },
      }
      wrapper.setState({ zoom: 1.1 })

      wrapper.find('.decrement').simulate('click')

      // then
      expect(wrapper.state('zoom')).toBe(1)
    })

    it('should increment zoom on click on plus button', () => {
      // given
      const getAttribute = jest.fn().mockImplementation(attribute => {
        if (attribute === 'step') {
          return 0.01
        }
        if (attribute === 'min') {
          return 1
        }
        if (attribute === 'max') {
          return 4
        }
      })

      // when
      const wrapper = shallow(<UploadThumb {...props} />)
      wrapper.instance()['handleSetZoomInput'] = {
        current: {
          getAttribute,
        },
      }

      wrapper.find('.increment').simulate('click')

      // then
      expect(wrapper.state('zoom')).toBe(1.1)
    })
  })

  describe('handleOnZoomChange', () => {
    it('should increase image size with value given by cursor change', () => {
      // given
      props.image = {
        name: 'IMG_4366.jpg',
        size: 1503804,
        type: 'image/jpeg',
      }

      // when
      const wrapper = shallow(<UploadThumb {...props} />)
      wrapper.find('input[name="zoomLeft"]').simulate('change', { target: { value: 1.95 } })

      // then
      expect(wrapper.state('zoom')).toBe(1.95)
    })
  })

  describe('handleOnImageChange', () => {
    const ctx = { fillStyle: '#000000' }
    const croppedRec = {
      height: 0.76,
      width: 0.45,
      x: 0.27,
      y: 0.11,
    }
    const getCroppingRect = jest.fn().mockImplementation(() => croppedRec)

    it('should not be called when no image', () => {
      // given
      props.image = null

      // when
      const wrapper = shallow(<UploadThumb {...props} />)

      wrapper.instance().handleOnImageChange(ctx)

      // then
      expect(props.onImageChange).not.toHaveBeenCalledWith()
    })

    it('should be called with ctx when onImageChange func given and upload is disabled', () => {
      // given
      props.image = {
        name: 'IMG_4366.jpg',
        size: 1503804,
        type: 'image/jpeg',
      }

      // when
      const wrapper = shallow(<UploadThumb {...props} />)
      wrapper.setState({ image, isUploadDisabled: true, size: 15038040 })
      wrapper.instance()['avatarEditor'] = {
        current: {
          getCroppingRect,
        },
      }
      wrapper.instance().handleOnImageChange(ctx)

      // then
      expect(props.onImageChange).toHaveBeenCalledWith(ctx)
    })

    it('should be called with image, ctx and new coordonnates', () => {
      // given
      const wrapper = shallow(<UploadThumb {...props} />)

      // when
      wrapper.setState({ image, isUploadDisabled: false, size: 15038040 })
      wrapper.instance()['avatarEditor'] = {
        current: {
          getCroppingRect,
        },
      }
      const currentImage = wrapper.state('image')
      wrapper.instance().handleOnImageChange(ctx)

      // then
      expect(props.onImageChange).toHaveBeenCalledWith(ctx, currentImage, croppedRec)
    })
  })

  describe('computeNewZoom', () => {
    it('should return a bigger zoom', () => {
      // given
      const { current, min, max, step, factor, direction } = defaultParams

      // when
      const newZoom = computeNewZoom(current, min, max, step, factor, direction)

      // then
      expect(newZoom).toBeGreaterThan(current)
    })

    it('should return a smaller zoom', () => {
      // given
      const { current, min, max, step, factor } = defaultParams
      const direction = -1

      // when
      const newZoom = computeNewZoom(current, min, max, step, factor, direction)

      // then
      expect(newZoom).toBeLessThan(current)
    })

    it('should not return zoom smaller than min', () => {
      // given
      const { min, max, step, factor } = defaultParams
      const current = min + step
      const direction = -1

      // when
      const newZoom = computeNewZoom(current, min, max, step, factor, direction)

      // then
      expect(newZoom).toStrictEqual(current)
    })

    it('should not return zoom greater than max', () => {
      // given
      const { min, max, step, factor, direction } = defaultParams
      const current = max - step

      // when
      const newZoom = computeNewZoom(current, min, max, step, factor, direction)

      // then
      expect(newZoom).toStrictEqual(current)
    })
  })

  describe('isImageTooLarge', () => {
    it('should be false is image not larger than allowed', () => {
      // when
      const result = isImageTooLarge(2.5)

      // then
      expect(result).toStrictEqual(false)
    })

    it('should be true is image is larger than allowed', () => {
      // when
      const result = isImageTooLarge(12.5)

      // then
      expect(result).toStrictEqual(true)
    })
  })
})
