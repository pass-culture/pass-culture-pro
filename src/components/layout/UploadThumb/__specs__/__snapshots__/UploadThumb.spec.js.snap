// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | UploadThumb | should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UploadThumb
    border={25}
    borderRadius={250}
    className={null}
    collectionName="fake"
    dispatch={[MockFunction]}
    hasExistingImage={false}
    height={250}
    image={
      Object {
        "name": "IMG_4366.jpg",
        "size": 1503804,
        "type": "image/jpeg",
      }
    }
    maxSize={10}
    onImageChange={[MockFunction]}
    storeKey="image"
    width={250}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="upload-thumb"
      >
        <div
          className="editor-zone has-image"
        >
          <d
            border={25}
            borderRadius={250}
            className="avatar editor"
            color={
              Array [
                255,
                255,
                255,
                0.6,
              ]
            }
            height={250}
            image={
              Object {
                "name": "IMG_4366.jpg",
                "size": 1503804,
                "type": "image/jpeg",
              }
            }
            onImageChange={[Function]}
            onImageReady={[Function]}
            onLoadFailure={[Function]}
            onLoadSuccess={[Function]}
            onMouseMove={[Function]}
            onMouseUp={[Function]}
            onPositionChange={[Function]}
            rotate={0}
            scale={1}
            style={Object {}}
            width={250}
          />
          <div
            id="zoomControl"
          >
            <button
              className="change-zoom decrement"
              onClick={[Function]}
              type="button"
            >
              <span>
                -
              </span>
            </button>
            <input
              className="zoom level-left"
              max="4"
              min="1"
              name="zoomLeft"
              onChange={[Function]}
              step="0.01"
              type="range"
              value={1}
            />
            <button
              className="change-zoom increment"
              onClick={[Function]}
              type="button"
            >
              <span>
                +
              </span>
            </button>
          </div>
        </div>
        <nav
          className="field content"
        />
      </div>,
      "className": "field",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="editor-zone has-image"
          >
            <d
              border={25}
              borderRadius={250}
              className="avatar editor"
              color={
                Array [
                  255,
                  255,
                  255,
                  0.6,
                ]
              }
              height={250}
              image={
                Object {
                  "name": "IMG_4366.jpg",
                  "size": 1503804,
                  "type": "image/jpeg",
                }
              }
              onImageChange={[Function]}
              onImageReady={[Function]}
              onLoadFailure={[Function]}
              onLoadSuccess={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              onPositionChange={[Function]}
              rotate={0}
              scale={1}
              style={Object {}}
              width={250}
            />
            <div
              id="zoomControl"
            >
              <button
                className="change-zoom decrement"
                onClick={[Function]}
                type="button"
              >
                <span>
                  -
                </span>
              </button>
              <input
                className="zoom level-left"
                max="4"
                min="1"
                name="zoomLeft"
                onChange={[Function]}
                step="0.01"
                type="range"
                value={1}
              />
              <button
                className="change-zoom increment"
                onClick={[Function]}
                type="button"
              >
                <span>
                  +
                </span>
              </button>
            </div>
          </div>,
          <nav
            className="field content"
          />,
        ],
        "className": "upload-thumb",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <d
                border={25}
                borderRadius={250}
                className="avatar editor"
                color={
                  Array [
                    255,
                    255,
                    255,
                    0.6,
                  ]
                }
                height={250}
                image={
                  Object {
                    "name": "IMG_4366.jpg",
                    "size": 1503804,
                    "type": "image/jpeg",
                  }
                }
                onImageChange={[Function]}
                onImageReady={[Function]}
                onLoadFailure={[Function]}
                onLoadSuccess={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                onPositionChange={[Function]}
                rotate={0}
                scale={1}
                style={Object {}}
                width={250}
              />,
              <div
                id="zoomControl"
              >
                <button
                  className="change-zoom decrement"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    -
                  </span>
                </button>
                <input
                  className="zoom level-left"
                  max="4"
                  min="1"
                  name="zoomLeft"
                  onChange={[Function]}
                  step="0.01"
                  type="range"
                  value={1}
                />
                <button
                  className="change-zoom increment"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    +
                  </span>
                </button>
              </div>,
            ],
            "className": "editor-zone has-image",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "border": 25,
                "borderRadius": 250,
                "className": "avatar editor",
                "color": Array [
                  255,
                  255,
                  255,
                  0.6,
                ],
                "height": 250,
                "image": Object {
                  "name": "IMG_4366.jpg",
                  "size": 1503804,
                  "type": "image/jpeg",
                },
                "onImageChange": [Function],
                "onImageReady": [Function],
                "onLoadFailure": [Function],
                "onLoadSuccess": [Function],
                "onMouseMove": [Function],
                "onMouseUp": [Function],
                "onPositionChange": [Function],
                "rotate": 0,
                "scale": 1,
                "style": Object {},
                "width": 250,
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <button
                    className="change-zoom decrement"
                    onClick={[Function]}
                    type="button"
                  >
                    <span>
                      -
                    </span>
                  </button>,
                  <input
                    className="zoom level-left"
                    max="4"
                    min="1"
                    name="zoomLeft"
                    onChange={[Function]}
                    step="0.01"
                    type="range"
                    value={1}
                  />,
                  <button
                    className="change-zoom increment"
                    onClick={[Function]}
                    type="button"
                  >
                    <span>
                      +
                    </span>
                  </button>,
                ],
                "id": "zoomControl",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span>
                      -
                    </span>,
                    "className": "change-zoom decrement",
                    "onClick": [Function],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "-",
                    },
                    "ref": null,
                    "rendered": "-",
                    "type": "span",
                  },
                  "type": "button",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "zoom level-left",
                    "max": "4",
                    "min": "1",
                    "name": "zoomLeft",
                    "onChange": [Function],
                    "step": "0.01",
                    "type": "range",
                    "value": 1,
                  },
                  "ref": Object {
                    "current": null,
                  },
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <span>
                      +
                    </span>,
                    "className": "change-zoom increment",
                    "onClick": [Function],
                    "type": "button",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "+",
                    },
                    "ref": null,
                    "rendered": "+",
                    "type": "span",
                  },
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "field content",
          },
          "ref": null,
          "rendered": false,
          "type": "nav",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="upload-thumb"
        >
          <div
            className="editor-zone has-image"
          >
            <d
              border={25}
              borderRadius={250}
              className="avatar editor"
              color={
                Array [
                  255,
                  255,
                  255,
                  0.6,
                ]
              }
              height={250}
              image={
                Object {
                  "name": "IMG_4366.jpg",
                  "size": 1503804,
                  "type": "image/jpeg",
                }
              }
              onImageChange={[Function]}
              onImageReady={[Function]}
              onLoadFailure={[Function]}
              onLoadSuccess={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              onPositionChange={[Function]}
              rotate={0}
              scale={1}
              style={Object {}}
              width={250}
            />
            <div
              id="zoomControl"
            >
              <button
                className="change-zoom decrement"
                onClick={[Function]}
                type="button"
              >
                <span>
                  -
                </span>
              </button>
              <input
                className="zoom level-left"
                max="4"
                min="1"
                name="zoomLeft"
                onChange={[Function]}
                step="0.01"
                type="range"
                value={1}
              />
              <button
                className="change-zoom increment"
                onClick={[Function]}
                type="button"
              >
                <span>
                  +
                </span>
              </button>
            </div>
          </div>
          <nav
            className="field content"
          />
        </div>,
        "className": "field",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="editor-zone has-image"
            >
              <d
                border={25}
                borderRadius={250}
                className="avatar editor"
                color={
                  Array [
                    255,
                    255,
                    255,
                    0.6,
                  ]
                }
                height={250}
                image={
                  Object {
                    "name": "IMG_4366.jpg",
                    "size": 1503804,
                    "type": "image/jpeg",
                  }
                }
                onImageChange={[Function]}
                onImageReady={[Function]}
                onLoadFailure={[Function]}
                onLoadSuccess={[Function]}
                onMouseMove={[Function]}
                onMouseUp={[Function]}
                onPositionChange={[Function]}
                rotate={0}
                scale={1}
                style={Object {}}
                width={250}
              />
              <div
                id="zoomControl"
              >
                <button
                  className="change-zoom decrement"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    -
                  </span>
                </button>
                <input
                  className="zoom level-left"
                  max="4"
                  min="1"
                  name="zoomLeft"
                  onChange={[Function]}
                  step="0.01"
                  type="range"
                  value={1}
                />
                <button
                  className="change-zoom increment"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    +
                  </span>
                </button>
              </div>
            </div>,
            <nav
              className="field content"
            />,
          ],
          "className": "upload-thumb",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <d
                  border={25}
                  borderRadius={250}
                  className="avatar editor"
                  color={
                    Array [
                      255,
                      255,
                      255,
                      0.6,
                    ]
                  }
                  height={250}
                  image={
                    Object {
                      "name": "IMG_4366.jpg",
                      "size": 1503804,
                      "type": "image/jpeg",
                    }
                  }
                  onImageChange={[Function]}
                  onImageReady={[Function]}
                  onLoadFailure={[Function]}
                  onLoadSuccess={[Function]}
                  onMouseMove={[Function]}
                  onMouseUp={[Function]}
                  onPositionChange={[Function]}
                  rotate={0}
                  scale={1}
                  style={Object {}}
                  width={250}
                />,
                <div
                  id="zoomControl"
                >
                  <button
                    className="change-zoom decrement"
                    onClick={[Function]}
                    type="button"
                  >
                    <span>
                      -
                    </span>
                  </button>
                  <input
                    className="zoom level-left"
                    max="4"
                    min="1"
                    name="zoomLeft"
                    onChange={[Function]}
                    step="0.01"
                    type="range"
                    value={1}
                  />
                  <button
                    className="change-zoom increment"
                    onClick={[Function]}
                    type="button"
                  >
                    <span>
                      +
                    </span>
                  </button>
                </div>,
              ],
              "className": "editor-zone has-image",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "border": 25,
                  "borderRadius": 250,
                  "className": "avatar editor",
                  "color": Array [
                    255,
                    255,
                    255,
                    0.6,
                  ],
                  "height": 250,
                  "image": Object {
                    "name": "IMG_4366.jpg",
                    "size": 1503804,
                    "type": "image/jpeg",
                  },
                  "onImageChange": [Function],
                  "onImageReady": [Function],
                  "onLoadFailure": [Function],
                  "onLoadSuccess": [Function],
                  "onMouseMove": [Function],
                  "onMouseUp": [Function],
                  "onPositionChange": [Function],
                  "rotate": 0,
                  "scale": 1,
                  "style": Object {},
                  "width": 250,
                },
                "ref": Object {
                  "current": null,
                },
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <button
                      className="change-zoom decrement"
                      onClick={[Function]}
                      type="button"
                    >
                      <span>
                        -
                      </span>
                    </button>,
                    <input
                      className="zoom level-left"
                      max="4"
                      min="1"
                      name="zoomLeft"
                      onChange={[Function]}
                      step="0.01"
                      type="range"
                      value={1}
                    />,
                    <button
                      className="change-zoom increment"
                      onClick={[Function]}
                      type="button"
                    >
                      <span>
                        +
                      </span>
                    </button>,
                  ],
                  "id": "zoomControl",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <span>
                        -
                      </span>,
                      "className": "change-zoom decrement",
                      "onClick": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "-",
                      },
                      "ref": null,
                      "rendered": "-",
                      "type": "span",
                    },
                    "type": "button",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "zoom level-left",
                      "max": "4",
                      "min": "1",
                      "name": "zoomLeft",
                      "onChange": [Function],
                      "step": "0.01",
                      "type": "range",
                      "value": 1,
                    },
                    "ref": Object {
                      "current": null,
                    },
                    "rendered": null,
                    "type": "input",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <span>
                        +
                      </span>,
                      "className": "change-zoom increment",
                      "onClick": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "+",
                      },
                      "ref": null,
                      "rendered": "+",
                      "type": "span",
                    },
                    "type": "button",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": false,
              "className": "field content",
            },
            "ref": null,
            "rendered": false,
            "type": "nav",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
